pkg/sdn/registry/netnamespace/rest.go

```diff
@@ -0,0 +1,218 @@
+	return rs.registry.ListNetNamespaces(ctx, label, field)
+}
+
+// Watch returns NetNamespaces events via a watch.Interface.
+// It implements rest.Watcher.
+func (rs *REST) Watch(ctx kapi.Context, label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error) {
+	return rs.registry.WatchNetNamespaces(ctx, label, field, resourceVersion)
+}
+
+func (*REST) New() runtime.Object {
+	return &api.NetNamespace{}
+}
+
+func (*REST) NewList() runtime.Object {
+	return &api.NetNamespaceList{}
+}
+
+func (rs *REST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object, bool, error) {
+	netns := obj.(*api.NetNamespace)
+	oldNetns, err := rs.registry.GetNetNamespace(ctx, netns.Name)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if errs := validation.ValidateNetNamespaceUpdate(oldNetns, netns); len(errs) > 0 {
+		return nil, false, errors.NewInvalid("netNamespace", netns.Name, errs)
+	}
+
+	createdNetID := false
+	changedNetID := true
+	if netns.NetID == nil {
+		err = rs.assignNetID(netns)
+		if err != nil {
+			return nil, false, err
+		}
+		createdNetID = true
+	} else if *oldNetns.NetID == *netns.NetID {
+		changedNetID = false
+	} else if *netns.NetID != util.GlobalVNID {
+		err = rs.vnids.Allocate(*netns.NetID)
+		if err != vnidallocator.ErrAllocated {
+			return nil, false, fmt.Errorf("NetID %d is not allocated, you can only use existing NetID during update")
+		}
+	}
+
+	out, err := rs.registry.UpdateNetNamespace(ctx, netns)
+	if err != nil {
+		if createdNetID {
+			er := rs.revokeNetID(netns)
+			if er != nil {
+				// these should be caught by an eventual reconciliation / restart
+				glog.Errorf("Error releasing netnamespace %s NetID %d: %v", netns.Name, *netns.NetID, er)
+			}
+		}
+		return nil, false, err
+	}
+
+	if changedNetID {
+		er := rs.revokeNetID(oldNetns)
+		if er != nil {
+			// these should be caught by an eventual reconciliation / restart
+			glog.Errorf("Error releasing netnamespace %s NetID %d: %v", oldNetns.Name, *oldNetns.NetID, er)
+		}
+	}
+	return out, false, nil
+}
+
+func (rs *REST) assignNetID(netns *api.NetNamespace) error {
+	if rs.isGlobalNamespace(netns) {
+		netns.NetID = new(uint)
+		*netns.NetID = util.GlobalVNID
+	} else if netns.NetID != nil {
+		// Try to respect the requested Net ID.
+		if err := rs.vnids.Allocate(*netns.NetID); err != nil {
+			el := fielderrors.ValidationErrorList{fielderrors.NewFieldInvalid("NetID", netns.NetID, err.Error())}
+			return errors.NewInvalid("NetNamespace", netns.Name, el)
+		}
+	} else {
+		// Allocate next available.
+		vnid, err := rs.vnids.AllocateNext()
+		if err != nil {
+			el := fielderrors.ValidationErrorList{fielderrors.NewFieldInvalid("NetID", netns.NetID, err.Error())}
+			return errors.NewInvalid("NetNamespace", netns.Name, el)
+		}
+		netns.NetID = &vnid
+	}
+	return nil
+
+}
+
+func (rs *REST) revokeNetID(netns *api.NetNamespace) error {
```

Comment:
Added NetNamespace cache to make it less expensive.
This cache is used in revokeNetID() and VNID repair controller.